core/src/main/java/com/linecorp/armeria/common/websocket/WebSocketCloseStatus.java

```diff
@@ -0,0 +1,361 @@
+        this.reasonText = checkNotNull(reasonText, "reasonText");
+    }
+
+    /**
+     * Returns the status code.
+     *
+     * @see <a href="https://datatracker.ietf.org/doc/html/rfc6455#section-7.4">Status Codes</a>
+     */
+    public int code() {
+        return statusCode;
+    }
+
+    /**
+     * Returns the text that indicates the reason for closure.
+     *
+     * @see <a href="https://datatracker.ietf.org/doc/html/rfc6455#section-7.4">Status Codes</a>
+     */
+    public String reasonText() {
+        return reasonText;
+    }
+
+    /**
+     * Equality of {@link WebSocketCloseStatus} only depends on {@link #code()}.
+     */
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (null == o || getClass() != o.getClass()) {
+            return false;
+        }
+
+        final WebSocketCloseStatus that = (WebSocketCloseStatus) o;
+
+        return statusCode == that.statusCode;
+    }
+
+    @Override
+    public int hashCode() {
+        return statusCode;
+    }
+
+    @Override
+    public String toString() {
+        String text = this.text;
+        if (text == null) {
+            // E.g.: "1000 Bye", "1009 Message too big"
+            this.text = text = code() + " " + reasonText();
+        }
+        return text;
+    }
+
+    /**
+     * Tells whether the {@code code} is valid.
+     */
+    public static boolean isValidStatusCode(int code) {
+        return code < 0 ||
+            1000 <= code && code <= 1003 ||
+            1007 <= code && code <= 1014 ||
+            3000 <= code;
+    }
+
+    /**
+     * Returns a {@link WebSocketCloseStatus} with the {@code code}. If the {@code code} is not predefined,
+     * a new instance is created.
+     */
+    public static WebSocketCloseStatus valueOf(int code) {
+        switch (code) {
+            case 1000:
+                return NORMAL_CLOSURE;
+            case 1001:
+                return ENDPOINT_UNAVAILABLE;
+            case 1002:
+                return PROTOCOL_ERROR;
+            case 1003:
+                return INVALID_MESSAGE_TYPE;
+            case 1005:
+                return EMPTY;
+            case 1006:
+                return ABNORMAL_CLOSURE;
+            case 1007:
+                return INVALID_PAYLOAD_DATA;
+            case 1008:
+                return POLICY_VIOLATION;
+            case 1009:
+                return MESSAGE_TOO_BIG;
+            case 1010:
+                return MANDATORY_EXTENSION;
+            case 1011:
+                return INTERNAL_SERVER_ERROR;
+            case 1012:
+                return SERVICE_RESTART;
+            case 1013:
+                return TRY_AGAIN_LATER;
+            case 1014:
+                return BAD_GATEWAY;
+            case 1015:
+                return TLS_HANDSHAKE_FAILED;
+            default:
+                return new WebSocketCloseStatus(code, "Close status #" + code);
```

Comment:
- How about caching?
- We should raise an exception if a user specified an invalid code.

Reply:
> We should raise an exception if a user specified an invalid code.

We do this in the constructor. Let me validate the status right away.

> How about caching?

Not sure if it's a good idea because:
- The close statuses that are mostly used are already defined as static variables.
- Close statuses are only used at the end of a WebSocket session which means it's rarely used.

It doesn't seem to have much of advantages, so I'm reluctant to cache them. 🤔